driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/internal/crypt/KeyManagementService.java

```diff
@@ -34,37 +35,41 @@
 import javax.net.ssl.SSLContext;
 import java.io.Closeable;
 import java.nio.channels.CompletionHandler;
+import java.security.NoSuchAlgorithmException;
 import java.util.List;
-import java.util.concurrent.TimeUnit;
+import java.util.Map;
 
 import static java.util.Collections.singletonList;
 import static java.util.concurrent.TimeUnit.MILLISECONDS;
 
 class KeyManagementService implements Closeable {
-    private final int defaultPort;
+    private final Map<String, SSLContext> kmsProviderSslContextMap;
+    private final int timeoutMillis;
     private final TlsChannelStreamFactoryFactory tlsChannelStreamFactoryFactory;
-    private final StreamFactory streamFactory;
 
-    KeyManagementService(final SSLContext sslContext, final int defaultPort, final int timeoutMillis) {
-        this.defaultPort = defaultPort;
+    KeyManagementService(final Map<String, SSLContext> kmsProviderSslContextMap, final int timeoutMillis) {
+        this.kmsProviderSslContextMap = kmsProviderSslContextMap;
         this.tlsChannelStreamFactoryFactory = new TlsChannelStreamFactoryFactory();
-        this.streamFactory = tlsChannelStreamFactoryFactory.create(SocketSettings.builder()
-                                                                           .connectTimeout(timeoutMillis, TimeUnit.MILLISECONDS)
-                                                                           .readTimeout(timeoutMillis, TimeUnit.MILLISECONDS)
-                                                                           .build(),
-                                                                   SslSettings.builder().enabled(true).context(sslContext).build());
+        this.timeoutMillis = timeoutMillis;
     }
 
     public void close() {
         tlsChannelStreamFactoryFactory.close();
     }
 
     Mono<Void> decryptKey(final MongoKeyDecryptor keyDecryptor) {
+        SocketSettings socketSettings = SocketSettings.builder()
+                .connectTimeout(timeoutMillis, MILLISECONDS)
+                .readTimeout(timeoutMillis, MILLISECONDS)
+                .build();
+        StreamFactory streamFactory = tlsChannelStreamFactoryFactory.create(socketSettings,
+                SslSettings.builder().enabled(true).context(kmsProviderSslContextMap.get(keyDecryptor.getKmsProvider())).build());
```

Comment:
This also doesn't mirror the sync KMS,  risks a NPE if no kmsProvider in the Map and there is no default context used.

Also there seems to be no logging in this version of the KMS but there is in sync?

Reply:
> This also doesn't mirror the sync KMS, risks a NPE if no kmsProvider in the Map and there is no default context used

I think it's ok because here we are using our SslSetting class, and null is allowed for the context.  In the Stream implementation, it will look for a null SSLContext, e.g. in `TlsChannelStreamFactoryFactory.TlsChannelStream#getSslContext`

> there seems to be no logging in this version of the KMS but there is in sync

Done